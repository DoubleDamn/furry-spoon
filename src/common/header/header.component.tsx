import * as React from "react";
import "./header.scss";
import { Button } from "@material-ui/core";
import { NavLink } from "react-router-dom";
import { CITIES_NAMES } from "../../typing/enam";

type P = {
  onClick(cityName: string): void;
};

export const Header: React.FC<P> = p => {
  const getData = (e: any) => {
    const name = e.currentTarget.id as keyof typeof CITIES_NAMES;
    p.onClick(`${CITIES_NAMES[name]}`);
  };

  return (
    <header className="header_container">
      <Button color="secondary" id="spb" onClick={getData}>
        <NavLink
          to={`/${CITIES_NAMES.spb}`}
          className="header_link"
          activeClassName="header_link--active"
          data-marker="menu-item"
        >
          Saint-Petersburg
        </NavLink>
      </Button>
      <Button color="secondary" id="rostov" onClick={getData}>
        <NavLink
          to={`/${CITIES_NAMES.rostov}`}
          className="header_link"
          activeClassName="header_link--active"
          data-marker="menu-item"
        >
          Rostov-On-Don
        </NavLink>
      </Button>
      <Button color="secondary" id="barcelona" onClick={getData}>
        <NavLink
          to={`/${CITIES_NAMES.barcelona}`}
          className="header_link"
          activeClassName="header_link--active"
          data-marker="menu-item"
        >
          Barcelona
        </NavLink>
      </Button>
    </header>
  );
};
