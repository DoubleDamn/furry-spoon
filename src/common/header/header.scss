.header_container {
  display: flex;
  align-items: center;
  background-color: #4e4d4a;
  height: 48px;
  box-shadow: 2px 2px 6px 0 rgba(0, 0, 0, 0.3);
  justify-content: center;

  .header_link {
    text-decoration: none;
    padding: 10px 5px;
    color: white;
    font-weight: 400;
    &--active {
      background-color: rgba(255, 150, 67, 0.57);
      border-radius: 5px;
    }
  }

  .header_logo {
    height: 90px;
    margin-right: auto;
  }
}
