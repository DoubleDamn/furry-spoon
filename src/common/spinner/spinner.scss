.spinner_loader {
	border: 5px solid #e96e50;
	border-radius: 50%;
	height: 30px;
	left: 50%;
	opacity: 0;
	position: absolute;
	top: 50%;
	width: 30px;

	animation: pulsate 1s ease-out;
	animation-iteration-count: infinite;
}

@keyframes pulsate {
	0% {
		opacity: 0;
		transform: scale(0.1);
	}
	50% {
		opacity: 1;
	}
	100% {
		opacity: 0;
		transform: scale(1.2);
	}
}
