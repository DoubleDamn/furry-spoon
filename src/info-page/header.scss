.header_container {
	display: flex;
	align-items: center;
	background-color: #d2d2d2;
	height: 48px;
	box-shadow: 2px 2px 6px 0 rgba(0, 0, 0, 0.3);
	padding: 0 20%;
}