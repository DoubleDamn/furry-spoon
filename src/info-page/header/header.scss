.header_container {
  display: flex;
  align-items: center;
  background-color: #d2d2d2;
  height: 48px;
  box-shadow: 2px 2px 6px 0 rgba(0, 0, 0, 0.3);
  padding: 0 20%;

  .header_link {
    text-decoration: none;
    padding: 3px;
    color: #3f51b5;
    &--active {
      background-color: rgba(155, 96, 255, 0.1);
      border-radius: 5px;
    }
  }
}
