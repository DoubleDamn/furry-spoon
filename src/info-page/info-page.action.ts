import axios from "axios";
import { WEATHER_ACTIONS } from "../typing/enum";

const API_KEY = "appid=28a658487aea99c56a8209ff5fbe7387";
export const BASE_URL = "https://api.openweathermap.org/data/2.5";

export type WeatherActionType = ReturnType<
  typeof IS_LOADING | typeof GET_WEATHER_DATA | typeof WEATHER_DATA_ERROR
>;

export const IS_LOADING = (cityName: string) =>
  <const>{
    type: WEATHER_ACTIONS.IS_LOADING,
    payload: cityName
  };

export const GET_WEATHER_DATA = (data: any) =>
  <const>{
    type: WEATHER_ACTIONS.GET_WEATHER_DATA,
    payload: data
  };

export const WEATHER_DATA_ERROR = (error: any) =>
  <const>{
    type: WEATHER_ACTIONS.WEATHER_DATA_ERROR,
    payload: error
  };

export const getWeatherForecast = (cityName: string) => {
  return axios.get(
    `${BASE_URL}/weather?q=${cityName}&${API_KEY}&units=metric`,
    {
      headers: { "Content-Type": "application/json" }
    }
  );
};
