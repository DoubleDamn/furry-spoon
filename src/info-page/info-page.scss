.widget_title {
	text-align: center;
	color: #e96e50;
	margin-bottom: 15px;
}

.widget_container {
	color: #4e4d4a;

	.widget_city-name {
		font: 500 24px/1.4 Oswald, Arial, sans-serif;
	}
}

.widget_table {
	font: 400 14px/1.4 Roboto, Arial, sans-serif;
	border: 1px solid #ddd;
	width: 80%;
	border-collapse: collapse;
	margin-top: 10px;

	tr:nth-child(odd) {
		background-color: #f9f9f9;
	}

	td {
		text-align: left;
		padding: 5px;
		border: 1px solid #ddd;
	}
}
