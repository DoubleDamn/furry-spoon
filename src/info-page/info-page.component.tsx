import * as React from 'react';
import { RouteComponentProps } from 'react-router';
import { Container } from '@material-ui/core';

import { Spinner } from '../common/spinner/spinner';
import { Header } from '../common/header/header.component';
import './info-page.scss';

type P = {
	isLoading: boolean;
	weatherHeader: Partial<WeatherInfo> & { temperature: number };
	widgetBody: WidgetBody[];
	errorMsg: string;

	getWeatherData(cityName: string): void;
} & RouteComponentProps;

export class InfoPage extends React.PureComponent<P> {
	public componentDidMount(): void {
		const cityName = this.props.match.url.slice(1);
		this.props.getWeatherData(cityName);
	}

	public render(): JSX.Element {
		if (this.props.errorMsg) {
		return <div>Sorry, something went wrong. Error: {this.props.errorMsg}</div>
		}
		return (
			<>
				<Header onClick={this.props.getWeatherData} />
				{this.props.isLoading ? (
					<Spinner />
				) : (
					<Container>
						<h2 className="widget_title">Current weather and forecasts in your city</h2>
						<div className="widget_container">
							{this.mainInfo}
							{this.widgetTable}
						</div>
					</Container>
				)}
			</>
		);
	}

	private get mainInfo(): JSX.Element {
		const { name, temperature, weather } = this.props.weatherHeader;
		return (
			<>
				<span className="widget_city-name">Weather in {name}</span>
				<div>
					<h3>Current temperature: {temperature} Â°C</h3>
				</div>
				<span>
					Cloudiness: {weather.main}, {weather.description}
				</span>
			</>
		);
	}

	private get widgetTable(): JSX.Element {
		return (
			<table className="widget_table">
				<tbody>
					{this.props.widgetBody.map((i, idk) => {
						return (
							<tr key={idk}>
								<td>{i.title}</td>
								<td>
									{i.value} {i.mark}
								</td>
							</tr>
						);
					})}
				</tbody>
			</table>
		);
	}
}
