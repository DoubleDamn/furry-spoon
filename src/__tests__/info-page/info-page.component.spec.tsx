import React from 'react';
import { shallow } from 'enzyme';

import { InfoPage } from '../../info-page/info-page.component';
import { CITIES_NAMES } from '../../typing/enum';
import { Spinner } from '../../common/spinner/spinner';
import { Container } from '@material-ui/core';
import { Header } from '../../common/header/header.component';

describe('InfoPage', () => {
	let component;
	const props = {
		isLoading: true,
		widgetBody: [],
		weatherHeader: {},
		match: { url: `/${CITIES_NAMES.spb}`, isExact: null, path: null, params: null },
		history: null,
		location: null,
		error: null,

		getWeatherData: jest.fn(),
	};

	beforeAll(() => {
		component = shallow(<InfoPage {...props} />);
	});

	afterEach(() => {
		jest.clearAllMocks();
	});

	it('should create component', () => {
		expect(component.instance()).toBeTruthy();
	});

	it('should call get data request when component mount', () => {
		const cityName = props.match.url.slice(1);

		component.instance().componentDidMount();
		expect(props.getWeatherData).toHaveBeenCalled();
		expect(props.getWeatherData).toHaveBeenCalledWith(cityName);
	});

	describe('WidgetPage Loader', () => {
		it('should match loading shapshot', () => {
			expect(component).toMatchSnapshot();
		});

		it('should display loader when data is loading', () => {
			expect(component.find(Spinner)).toHaveLength(1);
			expect(component.find(Container)).toHaveLength(0);
		});
	});

	describe('WidgetPage Widget', () => {
		const errorProps = {
			...props,
			isLoading: false,
			error: {message:'error message', code: 404},
		};

		beforeEach(() => {
			component = shallow(<InfoPage {...errorProps} />);
		});

		it('should match widget snapshot', () => {
			expect(component).toMatchSnapshot();
		});

		it('should error message', () => {
			expect(component.find(Header)).toHaveLength(0);
			expect(component.find(Spinner)).toHaveLength(0);
			expect(component.find(Container)).toHaveLength(0);
		});
	});
});
