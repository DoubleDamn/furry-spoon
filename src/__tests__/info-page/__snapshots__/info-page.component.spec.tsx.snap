// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InfoPage WidgetPage Loader should match loading shapshot 1`] = `ShallowWrapper {}`;

exports[`InfoPage WidgetPage Widget should match widget snapshot 1`] = `ShallowWrapper {}`;
